### DISCUSSION TITLE
(no issue found): Prepend header in xxxx_out_transmit function

### DISCUSSION THREAD

**Discussion by ZhiyuanYuanNJ** [ipxe/ipxe#1286] [Q&A]
In the source code of the USB network card, the `xxxx_out_transmit` function uses `iob_ensure_headroom` to ensure sufficient headroom, but after printing, this value keeps returning 0. Resulted in timeout of the chain instruction. Like the following functions:
```/**
 * Transmit packet
 *
 * @v dm96xx		DM96xx device
 * @v iobuf		I/O buffer
 * @ret rc		Return status code
 */
static int dm96xx_out_transmit ( struct dm96xx_device *dm96xx,
				 struct io_buffer *iobuf ) {
	struct dm96xx_tx_header *header;
	size_t len = iob_len ( iobuf );
	int rc;

	/* Prepend header */
	if ( ( rc = iob_ensure_headroom ( iobuf, sizeof ( *header ) ) ) != 0 )   //here
		return rc;
	header = iob_push ( iobuf, sizeof ( *header ) );
	header->len = cpu_to_le16 ( len );

	/* Enqueue I/O buffer */
	if ( ( rc = usb_stream ( &dm96xx->usbnet.out, iobuf, 0 ) ) != 0 )
		return rc;

	return 0;
}

**Comment by mcb30** [comment]
> In the source code of the USB network card, the `xxxx_out_transmit` function uses `iob_ensure_headroom` to ensure sufficient headroom, but after printing, this value keeps returning 0. Resulted in timeout of the chain instruction. Like the following functions:
> ```/**
>  * Transmit packet
>  *
>  * @v dm96xx		DM96xx device
>  * @v iobuf		I/O buffer
>  * @ret rc		Return status code
>  */
> static int dm96xx_out_transmit ( struct dm96xx_device *dm96xx,
> 				 struct io_buffer *iobuf ) {
> 	struct dm96xx_tx_header *header;
> 	size_t len = iob_len ( iobuf );
> 	int rc;
> 
> 	/* Prepend header */
> 	if ( ( rc = iob_ensure_headroom ( iobuf, sizeof ( *header ) ) ) != 0 )   //here
> 		return rc;
> ```

Not sure what problem you're trying to report here, sorry. A return status of zero indicates success.

**Reply by ZhiyuanYuanNJ** [reply]
Hi, `/* Prepend header */` frequent fails,  causing `chain http://192.168.50.2/boot.ipxe` to timeout and fail. This shouldn't be a server-side issue, remove this judgment, `chain ...` successful.

**Reply by mcb30** [reply]
Sorry, do you mean that you have found and fixed the problem, and it turned out to be completely unrelated to this code?

**Reply by ZhiyuanYuanNJ** [reply]
I just reported discovering this problem, and I'm not sure if it's appropriate to remove this judgment.

**Reply by mcb30** [reply]
OK, but no problem has yet been reported.  I'm going to close this discussion so that it doesn't cause confusion.  Please feel free to open a new discussion or issue if you find an actual problem.
